core/chain_indexer.go

```diff
@@ -370,14 +377,109 @@ func (c *ChainIndexer) indexerLoop(currentHeader *types.WorkObject, qiIndexerCh
 }
 
 func (c *ChainIndexer) PruneOldBlockData(blockHeight uint64) {
+	c.pruneLock.Lock()
 	blockHash := rawdb.ReadCanonicalHash(c.chainDb, blockHeight)
+	if rawdb.ReadAlreadyPruned(c.chainDb, blockHash) {
+		return
+	}
+	rawdb.WriteAlreadyPruned(c.chainDb, blockHash) // Pruning can only happen once per block
+	c.pruneLock.Unlock()
+
 	rawdb.DeleteInboundEtxs(c.chainDb, blockHash)
 	rawdb.DeletePendingEtxs(c.chainDb, blockHash)
 	rawdb.DeletePendingEtxsRollup(c.chainDb, blockHash)
 	rawdb.DeleteManifest(c.chainDb, blockHash)
 	rawdb.DeletePbCacheBody(c.chainDb, blockHash)
-	rawdb.DeleteSpentUTXOs(c.chainDb, blockHash)
+	createdUtxos, _ := rawdb.ReadCreatedUTXOKeys(c.chainDb, blockHash)
+	if len(createdUtxos) > 0 {
+		createdUtxosToKeep := make([][]byte, 0, len(createdUtxos)/2)
+		for _, key := range createdUtxos {
+			if len(key) == rawdb.UtxoKeyWithDenominationLength {
+				if key[len(key)-1] > types.MaxTrimDenomination {
+					// Don't keep it if the denomination is not trimmed
+					// The keys are sorted in order of denomination, so we can break here
+					break
+				}
+				key[rawdb.PrunedUtxoKeyWithDenominationLength+len(rawdb.UtxoPrefix)-1] = key[len(key)-1] // place the denomination at the end of the pruned key (11th byte will become 9th byte)
+			}
+			// Reduce key size to 9 bytes and cut off the prefix
+			key = key[len(rawdb.UtxoPrefix) : rawdb.PrunedUtxoKeyWithDenominationLength+len(rawdb.UtxoPrefix)]
+			createdUtxosToKeep = append(createdUtxosToKeep, key)
+		}
+		c.logger.Infof("Removed %d utxo keys from block %d", len(createdUtxos)-len(createdUtxosToKeep), blockHeight)
+		rawdb.WritePrunedUTXOKeys(c.chainDb, blockHeight, createdUtxosToKeep)
+	}
 	rawdb.DeleteCreatedUTXOKeys(c.chainDb, blockHash)
+	/*tutxos, _ := rawdb.ReadTrimmedUTXOs(c.chainDb, blockHash)
+	sutxos, err := rawdb.ReadSpentUTXOs(c.chainDb, blockHash)
+	if err != nil {
+		c.logger.Error("Failed to read spent utxos", "err", err)
+	} else {
+		sutxos = append(sutxos, tutxos...)
+		if len(sutxos) > 0 {
+			select {
+			case c.utxoKeyPrunerChan <- sutxos:
+			default:
+				c.logger.Warn("utxoKeyPrunerChan is full, dropping spent utxos")
+			}
+		}
+	}*/
+	rawdb.DeleteSpentUTXOs(c.chainDb, blockHash)
+	rawdb.DeleteTrimmedUTXOs(c.chainDb, blockHash)
+	rawdb.DeleteTrimDepths(c.chainDb, blockHash)
+	rawdb.DeleteCollidingKeys(c.chainDb, blockHash)
+}
+
+func (c *ChainIndexer) UTXOKeyPruner() {
```

Comment:
need a defer panic catch, since its a go routine, saw a node crash on the garden test, might be this go routine, we have catches everywhere else

Reply:
This function is never called. I will delete it

Reply:
Deleted